exploration/number-selection.md

```diff
@@ -92,44 +95,274 @@ ICU MF1 messages using `plural` and `selectordinal` should be representable in M
 
 ## Proposed Design
 
-Given that we already have a `:number`,
-it makes sense to add a `<matchSignature>` to it with an option
-
-```xml
-<option name="select" values="plural ordinal exact" default="plural" />
-```
-
-The default `plural` value is presumed to be the most common use case,
-and it affords the least bad fallback when used incorrectly:
-Using "plural" for "exact" still selects exactly matching cases,
-whereas using "exact" for "plural" will not select LDML category matches.
-This might not be noticeable in the source language,
+### Number Selection
+
+Number selection has three modes:
+- `exact` selection matches the operand to explicit numeric keys exactly
+- `plural` selection matches the operand to explicit numeric keys exactly
+  or to plural rule categories if there is no explicit match
+- `ordinal` selection matches the operand to explicit numeric keys exactly
+  or to ordinal rule categories if there is no explicit match
+
+
+### Functions
+
+The following functions use numeric selection:
+
+The function `:number` is the default selector for numeric values.
+
+The function `:plural` operates identically to `:number`, except:
+- the `select` type of this selector is always `plural`
+- the function is **not** valid as a formatter
+
+The function `:ordinal` operates identically to `:number`, except:
+- the `select` type of this selector is always `ordinal`
+  (This function is valid as a formatter)
+
+The function `:integer` operates identically to `:number`, except:
+- the `minimumFractionDigits` and `maximumFractionDigits` options
+  default to `0` for all values
+- the _options_ `minimumFractionDigits` and `maximumFractionDigits`
+  are invalid.
+
+### Operands
+
+The operand of a number selector is either a literal that matches the `number-literal`
+production in the [ABNF](/main/spec/message.abnf) and which is parsed by the 
+implementation into a number; or an implementation-defined numeric type; 
+or is some serialization of a number that the implementation recognizes.
+
+> For example, in Java, any subclass of `java.lang.Number` plus the primitive
+> types (`byte`, `short`, `int`, `long`, `float`, `double`, etc.) 
+> might be considered as the "implementation-defined numeric types".
+> Implementations in other programming languages would define different types
+> or classes according to their local needs.
+
+Implementations MUST treat literals that do not match `number-literal` as
+non-numeric string literals.
+Non-numeric values, including non-numeric string literals, result in
+a _Selection Error_.
+
+> _Ed note:_ Alternatively we can require that non-numeric values result in `*`
```

Comment:
I'll buy that. 

I don't think this can be a _Resolution Error_, since the value resolves correctly. It is the number function that is not accepting the input. In the case of a selector, this means that it can produce a _Selection Error_ (to indicate the type mismatch). In the case of a formatter, this means that it can produce a _Formatting Error_. If you read the list of _Resolution Errors_, you'll see that none of them fit this case. And I do think that it is _internal_ to `:number` (and friends) whether to accept a given value.

The behavior of the selector is as you suggest, with the `*` case being the pattern selected. If the value is used in the pattern, that value will use its fallback resolution. Either way there will be an error available to be emitted (in an implementation-defined way).

Reply:
While the operand value does resolve correctly in this case, it's the resolution of the expression that fails. Consider this nonsensical but formally valid message, for instance:

```
.local $x = {horse :number}
.match {$x}
1 {{I'm a cow}}
* {{Herding {$x} along}}
```

Here, we need to emit an error when resolving the `{horse :number}` expression, and then when using its value for selection. That same order of errors should be retained if the selector expression itself fails to resolve.

> If you read the list of _Resolution Errors_, you'll see that none of them fit this case.

Good point. The specific kind of resolution error isn't currently defined: https://github.com/unicode-org/message-format-wg/blob/e588afba76b0bc31251a96c2760c4905618ed733/spec/formatting.md?plain=1#L240-L241

Reply:
@eemeli Thanks for this.

I think your example is a good one, since we want the expression to fail in the `.local` rather than waiting for the `.match` or the formatting in the pattern.

Regarding the quoted lines, I'm not sure, having thought about it, what an "invalid value" is (we don't define a valid value). `null` might be an acceptable value for some kinds of calls? Ultimately, though I think we need a specific error type (so we can test for it, if nothing else). I think something like `Invalid Expression`, one subtype of which would be `Type Mismatch`.

I'll create an issue to track and a PR (if I have time) to propose it.